lib/active_merchant/billing/gateways/elavon.rb

```diff
@@ -366,30 +366,48 @@ def add_line_items(xml, level_3_data)
         }
       end
 
-      def add_stored_credential(xml, options)
+      def add_stored_credential(xml, payment_method, options)
+        return unless options[:stored_credential]
+
         network_transaction_id = options.dig(:stored_credential, :network_transaction_id)
-        case
-        when network_transaction_id.nil?
-          return
-        when network_transaction_id.to_s.include?('|')
-          oar_data, ps2000_data = options[:stored_credential][:network_transaction_id].split('|')
-          xml.ssl_oar_data oar_data unless oar_data.nil? || oar_data.empty?
-          xml.ssl_ps2000_data ps2000_data unless ps2000_data.nil? || ps2000_data.empty?
-        when network_transaction_id.to_s.length > 22
-          xml.ssl_oar_data options.dig(:stored_credential, :network_transaction_id)
-        else
-          xml.ssl_ps2000_data options.dig(:stored_credential, :network_transaction_id)
+        xml.ssl_recurring_flag recurring_flag(options) if recurring_flag(options)
+        xml.ssl_par_value options[:par_value] if options[:par_value]
+        xml.ssl_association_token_data options[:association_token_data] if options[:association_token_data]
+
+        unless payment_method.is_a?(String) || options[:ssl_token].present?
+          xml.ssl_approval_code options[:approval_code] if options[:approval_code]
+          if network_transaction_id.to_s.include?('|')
+            oar_data, ps2000_data = network_transaction_id.split('|')
+            xml.ssl_oar_data oar_data unless oar_data.blank?
+            xml.ssl_ps2000_data ps2000_data unless ps2000_data.blank?
+          elsif network_transaction_id.to_s.length > 22
+            xml.ssl_oar_data network_transaction_id
+          elsif network_transaction_id.present?
+            xml.ssl_ps2000_data network_transaction_id
+          end
         end
       end
 
+      def recurring_flag(options)
+        return unless reason = options.dig(:stored_credential, :reason_type)
+        return 1 if reason == 'recurring'
+        return 2 if reason == 'installment'
+      end
+
       def merchant_initiated_unscheduled(options)
         return options[:merchant_initiated_unscheduled] if options[:merchant_initiated_unscheduled]
-        return 'Y' if options.dig(:stored_credential, :initiator) == 'merchant' && options.dig(:stored_credential, :reason_type) == 'unscheduled' || options.dig(:stored_credential, :reason_type) == 'recurring'
+        return 'Y' if options.dig(:stored_credential, :initiator) == 'merchant' && options.dig(:stored_credential, :reason_type) == 'unscheduled'
+      end
+
+      def add_installment_fields(xml, options)
+        xml.ssl_payment_number options[:payment_number]
+        xml.ssl_payment_count options[:installments]
```

Comment:
These are new fields that we don't send today right? Will they for sure be present when we are trying to add them? 

I guess my overall question is should adding them here be conditional on whether they are present like 

```suggestion
        xml.ssl_payment_count options[:installments] if options[:installments]
```

Reply:
We won't merge these changes until these fields are being passed for Installment transactions are being performed. They will start to fail since these are required fields.

Reply:
So will these fields need to be passed for all customers before merging?

Reply:
Just customers who pass `reason_type: 'installment'` which is only 2 and we've communicated with them. I'm waiting for confirmation and I'll check on our side and then I'll merge.